examples/speculative_decoding/collect_hidden_states/send_conversations_for_hiddens.py

```diff
@@ -0,0 +1,217 @@
+        type=int,
+        default=3072,
+        help="""Maximum number of tokens in a conversation. Longer conversations will be skipped.
+        Defaults to 3072 tokens.""",
+    )
+    parser.add_argument(
+        "--input-file",
+        type=Path,
+        required=True,
+        help="""Path to the input `jsonl` file containing conversations.
+        Each entry must have a unique `conversation_id` field and a `conversations` field
+        containing a list of messages.""",
+    )
+    parser.add_argument(
+        "--output-dir",
+        type=Path,
+        required=True,
+        help="""Root directory in which to save the hidden states.
+        The data will be saved as a torch (`.pt`) dump file for each conversation.""",
+    )
+    parser.add_argument(
+        "--debug-max-num-conversations",
+        type=int,
+        default=None,
+        help="""For debugging purposes, limit the number of conversations processed.
+        Default is None, meaning no limit.""",
+    )
+
+    return parser.parse_args()
+
+
+async def main(args: argparse.Namespace) -> None:
+    all_conversations = []
+    with args.input_file.open("r", encoding="utf-8") as f:
+        all_conversations.extend([json.loads(line) for line in f if line.strip()])
+
+    if any(not entry.get("conversation_id") for entry in all_conversations):
+        msg = "All conversations must have a 'conversation_id' field."
+        raise ValueError(msg)
+
+    print("Loaded", len(all_conversations), "conversations from", args.input_file)
+
+    client: AsyncOpenAI = AsyncOpenAI(
+        api_key=args.openai_api_key,
+        base_url=args.base_url,
+    )
+
+    tokenizer = AutoTokenizer.from_pretrained(args.model)
+    if tokenizer.pad_token is None:
+        tokenizer.pad_token = tokenizer.eos_token
+    bos_token_id = tokenizer.bos_token_id
+    if bos_token_id is None:
+        raise ValueError("The tokenizer does not have a BOS token.")
+
+    temp_meta_file = Path("/tmp/meta.json")
+    if temp_meta_file.exists():
+        print(f"Temporary meta file {temp_meta_file} found, removing it.")
+        temp_meta_file.unlink()
+
+    output_dir = args.output_dir
+    output_dir.mkdir(parents=True, exist_ok=True)
+    num_invalid = 0
+    num_error = 0
+    num_too_long = 0
+    num_success = 0
+    num_total_conversations = min(
+        len(all_conversations), args.debug_max_num_conversations or len(all_conversations)
+    )
+    for entry in tqdm(
+        all_conversations[: args.debug_max_num_conversations],
+        desc="Processing conversations",
+        total=num_total_conversations,
+    ):
+        conversation_id = entry["conversation_id"]
+        conversations = entry["conversations"]
+        if not conversations or not isinstance(conversations, list):
+            num_invalid += 1
+            continue
+
+        hidden_states_file = output_dir / f"{conversation_id}.pt"
+
+        # Use /tmp/meta.json to communicate with the local serving engine.
```

Comment:
What modification do we need to the serving engine in order to run this script?

Reply:
https://github.com/vllm-project/vllm/compare/main...CentML:vllm:eagle3-prepare-train-data
needs a bit of polish but that's pretty much it